src/protocol/mod.rs

```diff
@@ -178,9 +200,19 @@ impl WriteState {
             Some(WriteSize { ref mut size, ref mut payload }) => {
                 match size.try_write(socket) {
                     Ok(WriterResult::Done) => {
-                        debug!("WriteSize {:?}: Transitioning to writing payload", token);
-                        let payload = mem::replace(payload, Writer::empty());
-                        Some(Some(WriteState::WriteData(payload)))
+                        let payload = mem::replace(payload, None);
+                        if let Some(payload) = payload {
+                            debug!("WriteSize {:?}: Transitioning to writing payload", token);
+                            Some(Some(WriteState::WriteData(payload)))
+                        } else {
+                            debug!("WriteSize {:?}: payload is empty. Done writing.", token);
+                            if outbound.is_empty() {
```

Comment:
Would it be better to put this below the match? Something like

```rust
if let Some(None) = update {
  if outbound.is_empty() {
    interest.remove(EventSet::writable());
  }
  interest.insert(EventSet::readable());
}
```

If you feel strongly about it, I don't mind leaving it the way it is. It just seems less repetitive and harder to mess up this way.